*** gcc/config/arm/t-arm-elf.orig	2012-10-14 21:39:11.832632246 +0200
--- gcc/config/arm/t-arm-elf	2012-10-14 21:43:23.715783593 +0200
***************
*** 17,26 ****
  # along with GCC; see the file COPYING3.  If not see
  # <http://www.gnu.org/licenses/>.
  
  MULTILIB_OPTIONS     = marm/mthumb
  MULTILIB_DIRNAMES    = arm thumb
! MULTILIB_EXCEPTIONS  = 
! MULTILIB_MATCHES     =
  
  #MULTILIB_OPTIONS     += mcpu=fa526/mcpu=fa626/mcpu=fa606te/mcpu=fa626te/mcpu=fmp626/mcpu=fa726te
  #MULTILIB_DIRNAMES    += fa526 fa626 fa606te fa626te fmp626 fa726te
--- 17,27 ----
  # along with GCC; see the file COPYING3.  If not see
  # <http://www.gnu.org/licenses/>.
  
+ # We have to build 'arm' and 'thumb' libraries, it seems.
  MULTILIB_OPTIONS     = marm/mthumb
  MULTILIB_DIRNAMES    = arm thumb
! #MULTILIB_EXCEPTIONS  = 
! #MULTILIB_MATCHES     =
  
  #MULTILIB_OPTIONS     += mcpu=fa526/mcpu=fa626/mcpu=fa606te/mcpu=fa626te/mcpu=fmp626/mcpu=fa726te
  #MULTILIB_DIRNAMES    += fa526 fa626 fa606te fa626te fmp626 fa726te
***************
*** 39,47 ****
  # Not quite true.  We can support hard-vfp calling in Thumb2, but how do we
  # express that here?  Also, we really need architecture v5e or later
  # (mcrr etc).
! MULTILIB_OPTIONS       += mfloat-abi=hard
! MULTILIB_DIRNAMES      += fpu
! MULTILIB_EXCEPTIONS    += *mthumb/*mfloat-abi=hard*
  #MULTILIB_EXCEPTIONS    += *mcpu=fa526/*mfloat-abi=hard*
  #MULTILIB_EXCEPTIONS    += *mcpu=fa626/*mfloat-abi=hard*
  
--- 40,48 ----
  # Not quite true.  We can support hard-vfp calling in Thumb2, but how do we
  # express that here?  Also, we really need architecture v5e or later
  # (mcrr etc).
! #MULTILIB_OPTIONS       += mfloat-abi=hard
! #MULTILIB_DIRNAMES      += fpu
! #MULTILIB_EXCEPTIONS    += *mthumb/*mfloat-abi=hard*
  #MULTILIB_EXCEPTIONS    += *mcpu=fa526/*mfloat-abi=hard*
  #MULTILIB_EXCEPTIONS    += *mcpu=fa626/*mfloat-abi=hard*
  
***************
*** 89,91 ****
--- 90,118 ----
  # MULTILIB_MATCHES    += mcpu?arm7=mcpu?arm600
  # MULTILIB_MATCHES    += mcpu?arm7=mcpu?arm610
  # MULTILIB_MATCHES    += mcpu?arm7=mcpu?arm620
+ #
+ 
+ # Only we don't actually want any ARM libraries. Or
+ # vanilla thumb libraries.
+ MULTILIB_EXCEPTIONS = marm* mthumb
+ 
+ # Build with any one of arm7tdmi, M0, M3 or M4 support.
+ MULTILIB_OPTIONS += mcpu=arm7tdmi-s/mcpu=cortex-m0/mcpu=cortex-m3/mcpu=cortex-m4
+ MULTILIB_DIRNAMES += arm7tdmi-s cortex-m0 cortex-m3 cortex-m4
+ 
+ # These don't make any sense without thumb, because GCC likes to
+ # tell you that you have to supply another commandline argument
+ # rather than just setting it itself.
+ MULTILIB_EXCEPTIONS += mcpu=arm7tdmi-s* mcpu=cortex-m0* mcpu=cortex-m3* mcpu=cortex-m4*
+ 
+ # All this just to get mfloat-abi=hard mfpu=fpv4-sp-d16 only specified for the M4
+ MULTILIB_OPTIONS += mfloat-abi=hard mfpu=fpv4-sp-d16
+ MULTILIB_DIRNAMES += float-abi-hard fpuv4-sp-d16
+ MULTILIB_EXCEPTIONS += mfloat* mthumb/mfloat*
+ MULTILIB_EXCEPTIONS += mfpu* mthumb/mfpu*
+ MULTILIB_EXCEPTIONS += mthumb/mcpu=cortex-m4/mfloat-abi=hard
+ MULTILIB_EXCEPTIONS += mthumb/mcpu=cortex-m4/mfpu=fpv4-sp-d16
+ MULTILIB_EXCEPTIONS += *arm7tdmi-s*mfloat-abi* *arm7tdmi-s*mfpu*
+ MULTILIB_EXCEPTIONS += *cortex-m3*mfloat-abi* *cortex-m3*mfpu*
+ MULTILIB_EXCEPTIONS += *cortex-m0*mfloat-abi* *cortex-m0*mfpu*
+ 
